### PR TITLE
Make libcxx compile more things by adding missing includes

### PR DESCRIPTION
**Pull request by chuggafan**
This adds the sources for the __libcpp_atomic_* so we don't need to have those erroneous builtins anymore, what we should have is define the actual builtins used for the file (e.g. the __atomic_exchange_n, etc.) and define that we have atomics from that. This only affects the build to allow the libcxx branch to properly build everything except that __builtin_strcmp in <typeinfo> and that mutex instantiation in line 658 of locale.cpp

### CODE REVIEW

**Comment by LADSoft** [comment]
thanks.   Can you elaborate on what is wrong with __builtin_strcmp and line 658 of locale.cpp?   The sources for the library compile fine for me (using /_LIBCPP_BUILDING_LIBRARY on the command line).

**Comment by chuggafan** [comment]
When doing omake fullbuild /j:8

```
occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\cregex.o cregex.cpp
Error(232)    C:\OrangeC\src\clibs\cpp\libcxx\include\typeinfo(197):  Could not find a match for '__builtin_strcmp(char const *, char const *)'
note:         C:\OrangeC\src\clibs\cpp\libcxx\include\memory(3513):  Referenced in instantiation of 'std::__1::__shared_ptr_pointer<__empty_state<char>*, default_delete<__empty_state<char>>, allocator<__empty_state<char>>>::__get_deleter<(*)( const type_info&) const >'
```

```

occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\system_error.o system_error.cpp
Error(199)    C:\OrangeC\src\clibs\cpp\libcxx\include\tuple(232):  Reference initialization of type 'std::__1::{unique}::__fake_bind' cannot be bound to type 'std::__1::{unique}::__fake_bind'
note:         C:\OrangeC\src\clibs\cpp\libcxx\include\tuple(232):  Referenced in instantiation of 'std::__1::__tuple_leaf<unsigned int=0, __fake_bind&&, bool=0>::__tuple_leaf<__fake_bind, bool, >(__fake_bind&&)'
note:         C:\OrangeC\src\clibs\cpp\libcxx\include\tuple(387):  Referenced in instantiation of 'std::__1::__tuple_impl<__tuple_indices<...unsigned int=0>, __fake_bind&&>::__tuple_impl<..., unsigned int=0, __fake_bind&&, ..., ..., _Tl, _Tl, >(__tuple_indices<...unsigned int=0>, __tuple_types<_Tl&&>, __tuple_indices<...>, __tuple_types<..._Tp>, _Tp&&)'
note:         C:\OrangeC\src\clibs\cpp\libcxx\include\tuple(793):  Referenced in instantiation of 'std::__1::tuple<__fake_bind&&>::tuple<__fake_bind, bool=0, bool=0, >(__fake_bind&&)'
note:         C:\OrangeC\src\clibs\cpp\libcxx\include\mutex(669):  Referenced in instantiation of 'std::__1::call_once<__fake_bind, ..._Args>(once_flag&, __fake_bind&&, void)'
note:         locale.cpp(658):  In template instantiation started here

```

```

occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\iostream.o iostream.cpp
Error( 47)    C:\OrangeC\src\clibs\cpp\stdinc_libcxx\typeinfo(41):  Structured type 'std::bad_cast' previously defined
Error( 47)    C:\OrangeC\src\clibs\cpp\stdinc_libcxx\typeinfo(42):  Structured type 'std::bad_typeid' previously defined
Error locale.cpp(39): Could not open  <langinfo.h>  for input

```

occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\string.o string.cpp
Error (0): Could not open strstream.cpp for input
        occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\utility.o utility.cpp
Error (0): Could not open string.cpp for input
        occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\memory.o memory.cpp
Error (0): Could not open utility.cpp for input
        occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\variant.o variant.cpp
Error (0): Could not open memory.cpp for input
        occ /c /! /DSTD_NEWHANDLER -D__BUILDING_LSCRTL_DLL /D_LIBCPP_BUILDING_SYSTEM_ERROR /D_LIBCPP_BUILDING_LIBRARY -IC:\OrangeC\src\clibs\cpp\stdinc_libcxx -IC:\OrangeC\src\clibs\stdinc -IC:\OrangeC\src\clibs\cpp\libcxx\include -IC:\OrangeC\src\clibs\platform\win32\wininc -IC:\OrangeC\src\clibs\platform\win32\wininc\win32 -IC:\OrangeC\src\sqlite3 -oC:\OrangeC\src\clibs\object\dll\vector.o vector.cpp
Error (0): Could not open variant.cpp for input
Error (0): Could not open vector.cpp for input

```

The "Could not open for input" ones were confusing af, but __builtin_strcmp in the typeinfo header simply does not exist, as it's supposed to be a builtin, if you raw include it, the program will crash because it's meant to be defined via the compiler builtin method, now, as for doing the strcmp stuff, I don't think doing it via the current assembly mechanism is good, I think having a more optimizable IR version of it, lowered into an IR version, then lowered into the assembly version, a la clang and (presumably) gcc, as they use information such as the length of the strings, if the thing being compared against is aligned or not, etc. to compile the code. But that's a whole 'nother story and another investigation in how other compilers function.

Long story short is that we need to support __builtin_strcmp for typeinfo header support.

**Comment by LADSoft** [comment]
ok i hadn't gotten so far as to try building the compiler itself yet :smile:   Thanks you saved me a lot of head scratching with this PR!

I decided to compile the C++ programs in the tests directory before working on the compiler itself, but am making slow progress.     There are three or four open problems with that.    Right now I'm dealing with the 'functional' header, specifically '__invokable_r' doesn't compile right yet...     Good news is that after a lot of hair-pulling the 'list' header compiles and runs properly.    Some of the  templates are recursive and it caused some headaches :smile:

Only thing that reliably compiles right now is the RTL, although there were some recently introduced bugs that broke even that.   I've fixed them now though....   

the 'cannot open for input' errors are a cascade from some previous error that kept something from compiling, not sure why they happen exactly.   I suspect that the CWD got changed during the compile though which is why it cannot find the files...   which likely makes it some issue with omake...   I should look at that sometime I suppose...

**Comment by chuggafan** [comment]
Yhea I just naturally go and do "omake fullbuild /j:8", if you're doing the libcxx tests, do you have any automated way to do the libcxx tests on just one separate test thing on appveyor that'd be great so that they also get run and our changes never kill the library.

**Comment by LADSoft** [comment]
hm I wasn't explicitly doing the libcxx tests but rather the tests in the orangec\tests directory but yeah that is a good idea.    Will open an issue for it..
